### PR TITLE
fix: correct csrftoken promise issues

### PR DESCRIPTION
**Pull request by glasnt**
## Description
Addresses #195 

Fixes some issues introduced in #290, including 404 issue on `checkout/` (should be `checkout`), csrftoken returning `'[object Promise]`'. 

Future changes required to correct error message passthrough, but those will be included in test coverage updates

## Checklist
- [ ] Added steps to reproduce the changes in this pull request
- [ ] Added relevant testing in this pull request
- [ ] Please **merge** this PR for me once it is approved.

### CODE REVIEW

File: client/src/utils/fetch.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ const baseRequest = {
   credentials: 'include',
 };
 
-async function _getAPI(uri) {
+const _getAPI = async uri => {
   const { API_URL } = getConfig();
```
Comment (grayside) [inline client/src/utils/fetch.js:23]: question: Why is getConfig a function instead of having config.js directly supply the config object?

File: client/src/utils/fetch.js [hunk 2]
```diff
@@ -19,7 +19,7 @@ const baseRequest = {
   credentials: 'include',
 };
 
-async function _getAPI(uri) {
+const _getAPI = async uri => {
```
Comment (grayside) [inline client/src/utils/fetch.js:22]: suggestion: #readability while uri is correct, reading uri vs. url can be confusing and error prone. Might we rename "uri" to resource and improve clarity?

This suggestion seems out of scope.

File: client/src/utils/fetch.js [hunk 3]
```diff
@@ -67,18 +67,27 @@ async function _getAPI(uri) {
...
-    const csrfToken = _getAPI('csrf_token');
+    const token = await _getAPI('csrf_token');
 
     await fetch(url, {
       method: 'POST',
-      headers: { 'X-CSRFToken': csrfToken },
+      headers: { 'X-CSRFToken': token.csrfToken },
       ...baseRequest,
     });
     callback && callback(); // callbacks handle error message parsing directly.
```
Comment (grayside) [inline client/src/utils/fetch.js:93]: question: It seems like if callback is undefined, the UI would enter an inconsistent state. When would callback be undefined? Should an undefined callback instead prevent the buy API call from being sent?

This is likely outside current PR scope.

File: client/src/utils/fetch.js [hunk 4]
```diff
@@ -67,18 +67,27 @@ async function _getAPI(uri) {
...
 
-async function _postAPI(uri, callback) {
+export const buyProduct = async (productId, callback) => {
+  let uri = `products/${productId}/purchase/`;
   const { API_URL } = getConfig();
 
   let url = `${API_URL}/${uri}`;
   try {
-    const csrfToken = _getAPI('csrf_token');
+    const token = await _getAPI('csrf_token');
```
Comment (grayside) [inline client/src/utils/fetch.js:86]: discussion: My understanding from https://stackoverflow.com/a/38236296 is that errors from fetch are limited to request/network errors. If that is the case, 4xx and 5xx errors are not going to be thrown, which means we aren't handling a swathe of error cases here.

It's at the edge of scope here, but would like to understand if this is a problem we need to address.

File: client/src/utils/fetch.js [hunk 5]
```diff
@@ -120,11 +117,34 @@ export const getSiteConfig = async () => {
...
+      // Submit form payload and pass back csrf token
+      const response = await fetch(`${API_URL}/checkout`, {
+        method: 'POST',
+        headers: { 'X-CSRFToken': token.csrfToken },
+        body: JSON.stringify(payload),
+        ...baseRequest,
+      });
+      checkoutStatus = await response.json();
+    } catch (error) {
+      errors = [error];
```
Comment (grayside) [inline client/src/utils/fetch.js:138]: Repeat: My discussion item about 4xx and 5xx errors above.

File: client/src/utils/fetch.js (hunk 1 above)
Comment (glasnt) [reply client/src/utils/fetch.js:23]: Answer to most comments, including this one: This PR reverts to the state that `buyProduct` and `checkout` were as at https://github.com/GoogleCloudPlatform/avocano/blob/8afc2b99fd0be3c3e4a37277ef4447b9b7a38065/client/src/utils/fetch.js. This PR is designed to revert the changes back to this known "working" state. Not "good" state, as there's still issues with it (as noted with all your comments), but it's a functional state.

**Review by grayside — APPROVED** [review]
LGTM to move forward partial revert.

**Comment by glasnt** [comment]
All comments marked as resolved here, but re-opened as https://github.com/GoogleCloudPlatform/avocano/issues/339

File: client/src/utils/fetch.js (hunk 5 above)
Comment (glasnt) [reply client/src/utils/fetch.js:138]: https://github.com/GoogleCloudPlatform/avocano/issues/339

File: client/src/utils/fetch.js (hunk 4 above)
Comment (glasnt) [reply client/src/utils/fetch.js:86]: https://github.com/GoogleCloudPlatform/avocano/issues/339

File: client/src/utils/fetch.js (hunk 3 above)
Comment (glasnt) [reply client/src/utils/fetch.js:93]: https://github.com/GoogleCloudPlatform/avocano/issues/339

File: client/src/utils/fetch.js (hunk 2 above)
Comment (glasnt) [reply client/src/utils/fetch.js:22]: https://github.com/GoogleCloudPlatform/avocano/issues/339

File: client/src/utils/fetch.js (hunk 1 above)
Comment (glasnt) [reply client/src/utils/fetch.js:23]: https://github.com/GoogleCloudPlatform/avocano/issues/339
